qiskit/quantum_info/operators/dihedral/random.py

```diff
@@ -53,7 +53,8 @@ def random_cnotdihedral(num_qubits, seed=None):
         random_invertible_binary_matrix,
     )
 
-    linear = random_invertible_binary_matrix(num_qubits, seed=rng)
+    seed = rng.integers(100000, size=1, dtype=np.uint64)[0]
+    linear = random_invertible_binary_matrix(num_qubits, seed=seed).astype(int)
```

Comment:
This will always result in a copy. You could use `.astype(int, copy=False)` but it still would need to copy. Also I think the dtype is incorrect shouldn't this be int8 because that's what cnot dihedral is using: https://github.com/qiskit/qiskit/blob/main/qiskit/quantum_info/operators/dihedral/dihedral.py#L129.

That being said wouldn't it just make more sense to change all the binary matrices in rust to use `Array2<i8>` to match this? It'd be the same size as `Array2<bool>` but then you could do a no-copy return to cnotdihedral and anywhere else using these functions.

Reply:
The original motivation for rust binary matrices to be of type `Array2<bool>` is that python classes `Clifford` and `LinearFunction` internally store a matrix of booleans, and we wanted to avoid copying there. IMHO, the dihedral class is less important then either of the above (at least right now). However, I do agree that the python synthesize size needs to be cleaned up.

Reply:
I agree with Sasha. As follows form the function name, the aim is only to handle binary matrices (e.g. bool).
In the future, the plan is to port the synthesis code to rust, and so we can handle the other classes later.